docs/chapters/chapter12.md

```diff
@@ -0,0 +1,284 @@
+### Chapter 12
+
+Run `./scripts/skipTo 10` to reset your app to chapter 10 (yes we want to move to chapter 10, not 11 due to some complexity that redux introduces for the purposes of this course). Then you can code along with this guide. There is not currently a solution folder for this chapter.
+
+# Chapter 12: Native Modules
+
+In this chapter, we will learn how to create and use native modules in React Native. Your React Native application code may need to interact with native platform APIs that aren't provided by React Native or an existing library. You can write the integration code yourself using a Turbo Native Module.
+
+## Tasks for this section [code-a-long] -- Android & iOS
+
+Our app is a simple local storage app that allows users to store and retrieve data. We will create a Turbo Native Module to interact with the native platform's storage API.
+
+### 1. `define a typed JavaScript specification` using one of the most popular JavaScript type annotation languages: Flow or TypeScript
+
+React Native's Codegen tool generates platform-specific code for Android and iOS from a TypeScript or Flow specification, defining methods and data types for native and JavaScript runtime communication. A Turbo Native Module comprises the specification, native code, and Codegen-generated interfaces.
+
+```tsx
+// /specs/NativeLocalStorage.ts
+import type { TurboModule } from 'react-native'
+import { TurboModuleRegistry } from 'react-native'
+
+export interface Spec extends TurboModule {
+  setItem(value: string, key: string): void
+  getItem(key: string): string | null
+  removeItem(key: string): void
+  clear(): void
+}
+
+export default TurboModuleRegistry.getEnforcing<Spec>('NativeLocalStorage')
+```
+
+### 2. `configure your dependency management system to run Codegen`, which converts the specification into native language interfaces
+
+```json
+// /package.json
+"codegenConfig": {
+  "name": "NativeLocalStorageSpec", // This name will be used later to generate headers for us, so we can choose whatever we want.
+  "type": "modules", // module (modules), view (components), or all (a mix).
+  "jsSrcsDir": "/specs", // relative path to the folder
+  "android": {
+    "javaPackageName": "com.nativelocalstorage" // inform codegen about the package name
+  }
+}
+```
+
+#### a. Generate Android Code
+
+```bash
+cd android
+./gradlew generateCodegenArtifactsFromSchema
```

Comment:
Do we need to tell them to run this in Android Studio?

Reply:
nope, in the terminal works